### PR TITLE
feat: Add FiniteStateMachine

### PR DESCRIPTION
**Pull request by idan**
> Update: per @Hugos68 feedback, I converted to a class and renamed to `FiniteStateMachine`. Otherwise, no changes to the initial PR.

Adds a new `useFSM` rune that allows you to create minimalistic finite state machines.

Ported over from https://github.com/githubnext/tiny-svelte-fsm (which was already MIT, and I'm the author there). I made a couple of minor fixes along the way as well, mostly to enable people to call `.debounce` in actions and lifecycle handlers.

This PR includes:

✅ Docs with an interactive demo and examples
✅ Tests

Per our conversation in the Svelte Discord:

CleanShot 2024-07-02 at 16 41 57@2x

### CODE REVIEW

**Comment by idan** [comment]
A lil' demo video of the demo at the top of the docs:

https://github.com/user-attachments/assets/fae59bbf-9252-4df5-b9f8-7e8e726b4960

**Comment by Hugos68** [comment]
Could you capitalize the types in example code? It's a typescript convention.

**Review by TGlide — CHANGES_REQUESTED** [review]
Awesome work, thank you so much! Left a few comments :)

File: packages/runed/src/lib/utilities/FiniteStateMachine/FiniteStateMachine.svelte.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
+export class FiniteStateMachine<StatesT extends string, EventsT extends string> {
+	#current = $state() as StatesT;
```
[the remark this answers is not in the document]
Comment (idan) [reply packages/runed/src/lib/utilities/FiniteStateMachine/FiniteStateMachine.svelte.ts:2]: This actually doesn't work, because then current ends up with the type of `StatesT | undefined` which we don't want. @Hugos68 suggested the type assertion as a workaround and it's better than the alternative IMO!

[the remark this answers is not in the document]
Comment (idan) [reply packages/runed/src/lib/utilities/FiniteStateMachine/FiniteStateMachine.svelte.ts:2]: Sorry, for more background: I originally implemented this as a function that returned an object, but @Hugos68 recommended that I turn it into a class. But then, I have to declare the `current` member outside the initializer, and I don't have the `initial` value where I'm wrapping it in the `$state()` rune.

tl;dr this is kind of the cost of doing it as a class, seems like a reasonable tradeoff?

File: packages/runed/src/lib/utilities/FiniteStateMachine/FiniteStateMachine.svelte.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		"event" in meta &&
+		"args" in meta
+	);
+}
+
+export type FSMLifecycle = "_enter" | "_exit";
+
+// Actions are either a StateT (string)
+// or a function that optionally returns a StateT
+export type Action<StatesT> = StatesT | ((...args: unknown[]) => StatesT | void);
```
[the remark this answers is not in the document]
Comment (idan) [reply packages/runed/src/lib/utilities/FiniteStateMachine/FiniteStateMachine.svelte.ts:95]: Ah good catch! I did this on my dev branch on the original project but forgot to copy it over.

File: sites/docs/content/utilities/finite-state-machine.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+## Notes
+
+`FiniteStateMachine` is a loving rewrite of
+[kenkunz/svelte-fsm](https://github.com/kenkunz/svelte-fsm).
+
+FSMs are ideal for representing many different kinds of systems and interaction patterns.
+`FiniteStateMachine` is an intentionally minimalistic implementation. If you're looking for a more
+powerful FSM library, [statelyai/xstate](https://github.com/statelyai/xstate) is an excellent
+library with more features&thinsp;—&thinsp;and a steeper learning curve.
```
[the remark this answers is not in the document]
Comment (idan) [reply sites/docs/content/utilities/finite-state-machine.md:214]: Thank you! 🥰

**Comment by idan** [comment]
@TGlide thanks, what a lovely review.

I fixed all the things except for the generic vs type assertion. I don't have a better idea, but from a DX perspective, I definitely want the current state to be of type `StateT` not `StateT | undefined` or anything like that.

**Review by TGlide — APPROVED** [review]
Thank you again! LGTM 🎉
